Continuous integration: are pull requests bad? (#264)
mauroservienti/milestone.github.io: _posts/continuous-integration-are-pull-requests-bad.md

layout: post
header_image: /img/posts/continuous-integration-are-pull-requests-bad/header.jpg
title: "Continuous integration: are pull requests bad?"
author: Mauro Servienti
synopsis: "There are continuous integration advocates and pull requests evangelists. They both present valid arguments to sustain their thesis. Is there a different point of view we should be considering before choosing one style or the other?"
tags:
- processes
---

Continuous integration (CI) is the practice of continuously pushing code changes to the mainline instead of the established or accepted practice of going through pull requests. CI supporters assume that pull requests slow you down because the review process implies too much waiting time. When waiting time is too much, the assumption is that the people waiting for the code review will shift their attention to something else, first context switch issue, and when the review requested changes come in, they can go back to the pull request to fix them, second context switch issue. The conclusion is that pull requests are bad because they lead to too much context switching.

> Source: https://infrastructure-as-code.com/book/2021/01/02/pull-requests.html

> It might be the case in specific scenarios for sure, but I generally find it hard to point my finger at pull requests as the root cause of the problem. The practice of using pull requests can be for sure something that nudges in the direction of accepting long-wait time in reviews, but I fail to see the causation. If any, it's probably the other way around. When the organization structure pushes in the direction of code reviews that take a long time and is performed by a different group of people than the one committing the code, the only way to go through this frustrating process is through pull requests and the accompanying context switch. In fact, in most scenarios, this group is referred to as the committee. If this is the case, can we say that pull requests are the culprit? In essence, look for the cause, not the symptoms.

Reviewer: ```suggestion
It might be the case in specific scenarios for sure, but I generally find it hard to point my finger at pull requests as the root cause of the problem. The practice of using pull requests can be for sure something that nudges in the direction of accepting long wait times in reviews, but I fail to see the causation. If anything, it's probably the other way around. When the organization structure pushes in the direction of code reviews that takes a long time and is performed by a different group of people than the one committing the code; the only way to go through this frustrating process is through pull requests and the accompanying context switch. In fact, in most scenarios, this group is referred to as the committee. If this is the case, can we say that pull requests are the culprit? In essence, look for the cause, not the symptoms.
```

> I also feel that using the context switching argument as the main problem is a weak argument. It seems to assume that a development team is doing one thing at a time, and thus while waiting for a pull request review, the team has to find something else to do and switch the context. In any organization, many things can slow down anyone; Being assigned to more than one task at a time is a well-known and established practice that helps fill any slack time caused by interruptions in the flow. I also find it hard to believe that once the pull request is ready, a development team has nothing else to do in the same context or for the same task, like documentation. The group can easily organize themselves to fill the waiting time caused by the pull request review.

Reviewer: ```suggestion
I also feel that using the context switching argument as the main problem is a weak argument. It seems to assume that a development team is doing one thing at a time, and thus while waiting for a pull request review, the team has to find something else to do and switch the context. In any organization, many things can slow down anyone; being assigned to more than one task at a time is a well-known and established practice that helps fill any slack time caused by interruptions in the flow. I also find it hard to believe that once the pull request is ready, a development team has nothing else to do in the same context or for the same task, like documentation. The group can easily organize themselves to fill the waiting time caused by the pull request review.
```

However, let's assume that it's impossible.

## Poka-yokes

> Pull requests and the inherent need for reviews are a [poka-yoke](https://seths.blog/2021/11/in-search-of-poka-yokes/), and as such, we should treat them.

Reviewer: ```suggestion
Pull requests and the inherent need for reviews are a [poka-yoke](https://seths.blog/2021/11/in-search-of-poka-yokes/), and as such, we should treat them as such.
```

> Pull requests and the inherent need for reviews are a [poka-yoke](https://seths.blog/2021/11/in-search-of-poka-yokes/), and as such, we should treat them.

Reviewer: Might be worth a brief description of "poka-yoke" so people unfamiliar with the term don't need to click through to learn about it.
